dns_cache/manager.go

```diff
@@ -0,0 +1,62 @@
+package dns_cache
+
+import (
+	"context"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/TykTechnologies/tyk/log"
+)
+
+var (
+	logger = log.Get().WithField("prefix", "dns-cache")
+)
+
+type dialContextFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+type IDnsCacheManager interface {
+	InitDNSCaching(ttl, checkInterval time.Duration)
+	WrapDialer(dialer *net.Dialer) dialContextFunc
+}
+
+type DnsCacheManager struct {
+	DnsCache *DnsCacheStorage
+}
+
+func NewDnsCacheManager() *DnsCacheManager {
+	return &DnsCacheManager{nil}
+}
+
+func (m *DnsCacheManager) WrapDialer(dialer *net.Dialer) dialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		return m.doCachedDial(dialer, ctx, network, address)
+	}
+}
+
+func (m *DnsCacheManager) doCachedDial(d *net.Dialer, ctx context.Context, network, address string) (net.Conn, error) {
+	if m.DnsCache == nil {
+		return d.DialContext(ctx, network, address)
+	}
+
+	separator := strings.LastIndex(address, ":")
+	ips, err := m.DnsCache.FetchItem(address[:separator])
```

Comment:
seems like we break here idea behind DNS load balancing (multiple IPs). if we reach this function - when new connection is about to do dial-call - that means we need to hit actual DNS because it will give us IP list with different order so connection will be load-balanced to another IP behind that domain name. Probably the right solution would be - don't cache DNS calls if resolver returns list of IPs per domain name